### PR TITLE
Build Global tool for PowerShell and SDK container

### PR DESCRIPTION
**Pull request by adityapatwardhan**
# PR Summary
This PR builds a unified global tool to support all the runtimes as well as two separate reduced size global tools to be included in dotnet sdk container images.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [x] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [ ] [Add `[feature]` to your commit messages if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

### CODE REVIEW

File: src/Microsoft.PowerShell.GlobalTool.Shim/GlobalToolShim.cs [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+        private const string UnixFolderName = "unix";
+
+        /// <summary>
+        /// Entry point for the global tool.
+        /// </summary>
+        /// <param name="args">Arguments passed to the global tool.</param>
+        public static void Main(string[] args)
+        {
+            var currentPath = new FileInfo(System.Reflection.Assembly.GetEntryAssembly().Location).Directory.FullName;
+            var isWindows = Environment.OSVersion.Platform == System.PlatformID.Win32NT;
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.GlobalTool.Shim/GlobalToolShim.cs]: In CorePsPlatform.cs we have
```c#
        public static bool IsWindows
        {
            get
            {
                return RuntimeInformation.IsOSPlatform(OSPlatform.Windows);
            }
        }
```

File: src/Microsoft.PowerShell.GlobalTool.Shim/runtimeconfig.template.json [hunk 2]
```diff
@@ -0,0 +1,3 @@
+{
+  "rollForwardOnNoCandidateFx": 2
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.GlobalTool.Shim/runtimeconfig.template.json]: Could you add comment why we need this?

File: tools/packaging/packaging.psm1 [hunk 3]
```diff
@@ -3092,105 +3092,152 @@ function Get-PackageVersionAsMajorMinorBuildRevision
...
-        [Parameter(Mandatory)] $ReleaseTag,
+    if (-not (Test-Path $path))
+    {
+        throw "Path not found: $Path"
+    }
 
-        [Parameter(Mandatory)] $DestinationPath
-    )
+    ## Remove unnecessary files
+    $localeFolderToRemove = 'cs', 'de', 'es', 'fr', 'it', 'ja', 'ko', 'pl', 'pt-BR', 'ru', 'tr', 'zh-Hans', 'zh-Hant'
```
Comment (iSazonov) [inline tools/packaging/packaging.psm1]: What if there will new folder? Maybe not "remove" but "use/filter only necessary folder"?

File: src/Microsoft.PowerShell.GlobalTool.Shim/GlobalToolShim.cs [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+            var pwshPath = Path.Combine(currentPath, platformFolder, PwshDllName);
+            string processArgs = string.IsNullOrEmpty(argsString) ? $"{pwshPath}" : $"{pwshPath} -c {argsString}";
+
+            if (File.Exists(pwshPath))
+            {
+                System.Diagnostics.Process.Start("dotnet", processArgs).WaitForExit();
+            }
+            else
+            {
+                throw new FileNotFoundException(pwshPath);
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.GlobalTool.Shim/GlobalToolShim.cs]: Do we get really more useful exception message? Maybe not check File.Exist?

File: src/Microsoft.PowerShell.GlobalTool.Shim/GlobalToolShim.cs (hunk 1 above)
Comment (adityapatwardhan) [reply src/Microsoft.PowerShell.GlobalTool.Shim/GlobalToolShim.cs]: Did not want to create a dependency from this `dll` to `SMA.dll` for just one function.

Comment (iSazonov) [reply src/Microsoft.PowerShell.GlobalTool.Shim/GlobalToolShim.cs]: Sorry, I meant that we could use `RuntimeInformation.IsOSPlatform(OSPlatform.Windows)`.

File: src/Microsoft.PowerShell.GlobalTool.Shim/runtimeconfig.template.json (hunk 2 above)
Comment (adityapatwardhan) [reply src/Microsoft.PowerShell.GlobalTool.Shim/runtimeconfig.template.json]: Added

File: src/Microsoft.PowerShell.GlobalTool.Shim/GlobalToolShim.cs (hunk 1 above)
Comment (adityapatwardhan) [reply src/Microsoft.PowerShell.GlobalTool.Shim/GlobalToolShim.cs]: Changed to use `RuntimeInformation`

File: tools/packaging/packaging.psm1 (hunk 3 above)
Comment (adityapatwardhan) [reply tools/packaging/packaging.psm1]: Can I take this as a bug fix after this PR?

File: src/Microsoft.PowerShell.GlobalTool.Shim/GlobalToolShim.cs (hunk 4 above)
Comment (adityapatwardhan) [reply src/Microsoft.PowerShell.GlobalTool.Shim/GlobalToolShim.cs]: We want to convey that the file was not found. The error message by `FileNotFoundException` is sufficient. Something similar to:

```Powershell
PS> Write-Error -Exception $q
Write-Error -Exception $q : C:\doesnotexist
+ CategoryInfo          : NotSpecified: (:) [Write-Error], FileNotFoundException
+ FullyQualifiedErrorId : System.IO.FileNotFoundException
```

We need not add resource file if the default exception is ok.

File: tools/packaging/packaging.psm1 (hunk 3 above)
Comment (iSazonov) [reply tools/packaging/packaging.psm1]: Yes.

File: src/Microsoft.PowerShell.GlobalTool.Shim/GlobalToolShim.cs (hunk 4 above)
Comment (iSazonov) [reply src/Microsoft.PowerShell.GlobalTool.Shim/GlobalToolShim.cs]: My question was about:
```
> dotnet c:\rrrrrrr.dll
No executable found matching command "dotnet-c:\rrrrrrr.dll"
```

Comment (adityapatwardhan) [reply src/Microsoft.PowerShell.GlobalTool.Shim/GlobalToolShim.cs]: I would not call `Process.Start` if the `pwshPath` does not exist.

File: tools/packaging/packaging.strings.psd1 [hunk 5]
```diff
@@ -147,4 +147,37 @@ open {0}
...
+<?xml version="1.0" encoding="utf-8"?>
+<package xmlns="http://schemas.microsoft.com/packaging/2013/05/nuspec.xsd">
+    <metadata>
+        <id>{0}</id>
+        <version>{1}</version>
+        <authors>Microsoft</authors>
+        <owners>Microsoft,PowerShell</owners>
+        <projectUrl>https://github.com/PowerShell/PowerShell</projectUrl>
+        <iconUrl>https://github.com/PowerShell/PowerShell/blob/master/assets/Powershell_black_64.png?raw=true</iconUrl>
+        <requireLicenseAcceptance>true</requireLicenseAcceptance>
```
Comment (TravisEz13) [inline tools/packaging/packaging.strings.psd1]: This should be false

File: tools/releaseBuild/azureDevOps/templates/nuget.yml [hunk 6]
```diff
@@ -36,9 +38,9 @@ jobs:
   - task: DownloadBuildArtifacts@0
     displayName: 'Download PowerShell build artifacts'
     inputs:
-      buildType: specific
-      project: '9d7f7aa4-6f41-480d-a367-82eb278461a1'
-      pipeline: 696
+      buildType: current
+      #project: '9d7f7aa4-6f41-480d-a367-82eb278461a1'
```
Comment (TravisEz13) [inline tools/releaseBuild/azureDevOps/templates/nuget.yml]: remove comments

**Review by TravisEz13 — APPROVED** [review]

File: tools/packaging/packaging.strings.psd1 (hunk 5 above)
Comment (adityapatwardhan) [reply tools/packaging/packaging.strings.psd1]: Fixed

File: tools/releaseBuild/azureDevOps/templates/nuget.yml (hunk 6 above)
Comment (adityapatwardhan) [reply tools/releaseBuild/azureDevOps/templates/nuget.yml]: Fixed

**Pull request merged**